docs/proposals/201912_thanos_binary_index_header.md

```diff
@@ -0,0 +1,191 @@
+    * Download index-cache.json from bucket
+* Load whole index-cache.json to memory and keep it for lifetime of the block.
+
+The current, mentioned [index-cache.json][index-cache-code] holds block’s:
+* [TOC][toc]
+* [Symbols][symbols]
+* LabelValues:
+  * Calculated from all [LabelsIndicesTable][labels-indices-table], that each points to [LabelIndex][label-index] entry.
+  * Used for matching against request selectors. 
+* [Postings Offsets][postings-offsets]
+
+There are few problems with this approach: 
+
+1. Downloading the whole TSDB index file (which can be from few MB up to 64GB or even more when limit will be increased soon) is expensive and might be slow.
+2. Building such index-cache requires at least 2x `index-cache.json` size of memory.
+3. Parsing large index-cache.json files (which can be from few KB to few GB) might be slow.
+4. We keep the full content of index-cache file in memory for all blocks.
+   * Some blocks are never queried. 
+   * Some posting offsets and label values within each block are never used.
+
+1, 2 & 3 contributes to Store Gateway startup being slow and resource consuming: https://github.com/thanos-io/thanos/issues/448
+
+4. causes unnecessary cache for unused blocks.
+
+This design is trying to address those 4 problems. 
+
+## Goals
+    
+* Reduce confusion between index-cache.json and IndexCache for series and postings.
+* Do not keep large pieces of the blocks (e.g symbols, label values) in the memory for all blocks. Be able to load it quickly in query time from disk.
+* Use constant amount of memory for building index-header/index-cache; download only required pieces of index from bucket and compute new TOC file.
+* Leverage mmap and let OS to unload unused pages from the physical memory as needed to satisfy the queries against the block.
+ This is due to random access against label values and postings offsets.
+
+## No Goals
+
+* Full, no initial startup as designed in [Cortex, no initial block sync](https://github.com/thanos-io/thanos/issues/1813)
+  * However this proposal might a step towards that as we might be able to load index-cache/index quickly on demand from disk.
+  * In the same time be able to load `index-header` in query time directly from bucket is not a goal of this proposal.
+* Decreasing size. While it would be nice to use less space; our aim is latency of building/loading the block. That might be correlated with size, but not necessarily (e.g when extra considering compression)
+
+## Verification
+
+* Compare memory and latency usage for startup without index-header being generated on compactor
+* Benchmark for loading the `index-header` vs `index-cache.json` into Store GW.
+
+## Proposal
+
+TSDB index is in binary [format](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md). 
+
+To allow reduced resource consumption and effort when building (1), (2), "index-header" for blocks we plan to reuse similar format for sections like symbols, label indices and 
```

Comment:
Not sure if I have missed this detail but what is the actual format that this proposes? Protobufs?

Reply:
A custom binary format was planned, composed with those sections as mentioned, however, I might find a blocker to this ):